File: workers/research/src/black_book_research/adapters/federal/shared/parser.py
"""Parses federal adapter fixture exports into normalized candidate records (BB-046)."""

from __future__ import annotations

from black_book_research.adapters.candidates import stamp_candidate_provenance
from black_book_research.adapters.types import SourceRegistryEntry

from .export_filter import filter_large_export_payload
from .retention import partition_by_retention
from .types import FederalAdapterDefinition, FederalParseResult


def parse_federal_fixture_batch(
    definition: FederalAdapterDefinition,
    entry: SourceRegistryEntry,
    run_id: str,
    captured_at: str,
    raw: object,
) -> FederalParseResult:
    if not isinstance(raw, list):
        raise ValueError("Federal fixture batch must be an array")

    records: list[dict[str, object]] = []
    for index, item in enumerate(raw):
        if not isinstance(item, dict):
            raise ValueError(f"Federal fixture record at index {index} must be an object")
        records.append(item)

    qualified, rejected = partition_by_retention(records, definition.retention)
    candidates = []
    filtered_export_count = 0

    for record in qualified:
        stable_identifier = str(record.get("stableIdentifier") or record.get("id") or "").strip()
        filtered = filter_large_export_payload(record, definition.export_filter)
        if filtered.filtered:
            filtered_export_count += 1

        title = record.get("title")
        canonical_url = record.get("canonicalUrl")
        classification = record.get("classification")

        candidates.append(
            stamp_candidate_provenance(
                entry,
                run_id=run_id,
                captured_at=captured_at,
                stable_identifier=stable_identifier,
                title=title if isinstance(title, str) else None,
                canonical_url=canonical_url if isinstance(canonical_url, str) else None,
                classification=(
                    classification if isinstance(classification, str) else definition.contract.classification
                ),
                payload=filtered.payload,
            )
        )

    return FederalParseResult(
        candidates=tuple(candidates),
        rejected=rejected,
        filtered_export_count=filtered_export_count,
    )

File: packages/domain/src/adapters/federal/shared/parser.ts
/**
 * Parses federal adapter fixture exports into normalized candidate records.
 */
import { stampCandidateProvenance } from '../../candidates.js';
import type { SourceRegistryEntry } from '../../types.js';
import { filterLargeExportPayload } from './export-filter.js';
import { partitionByRetention } from './retention.js';
import type {
  FederalAdapterDefinition,
  FederalParseResult,
  RawFederalExportRecord,
} from './types.js';

function asExportRecords(raw: unknown): RawFederalExportRecord[] {
  if (!Array.isArray(raw)) {
    throw new Error('Federal fixture batch must be an array');
  }
  return raw.map((item, index) => {
    if (!item || typeof item !== 'object') {
      throw new Error(`Federal fixture record at index ${index} must be an object`);
    }
    return item as RawFederalExportRecord;
  });
}

export function parseFederalFixtureBatch(
  definition: FederalAdapterDefinition,
  entry: SourceRegistryEntry,
  runId: string,
  capturedAt: string,
  raw: unknown,
): FederalParseResult {
  const records = asExportRecords(raw);
  const { qualified, rejected } = partitionByRetention(records, definition.retention);

  const candidates = [];
  let filteredExportCount = 0;

  for (const record of qualified) {
    const stableIdentifier = String(record.stableIdentifier ?? record.id ?? '').trim();
    const filtered = filterLargeExportPayload(record, definition.exportFilter);
    if (filtered.filtered) {
      filteredExportCount += 1;
    }

    candidates.push(
      stampCandidateProvenance(entry, runId, capturedAt, {
        stableIdentifier,
        ...(typeof record.title === 'string' ? { title: record.title } : {}),
        ...(typeof record.canonicalUrl === 'string' ? { canonicalUrl: record.canonicalUrl } : {}),
        classification:
          typeof record.classification === 'string'
            ? record.classification
            : definition.contract.classification,
        payload: filtered.payload,
      }),
    );
  }

  return { candidates, rejected, filteredExportCount };
}
